### DISCUSSION TITLE
Only reflect attributes if attributes are set / if needed?

### DISCUSSION THREAD

**Discussion by schjetne** [lit/lit#4854] [Q&A] [answered]
I've been wondering about this since I started using Lit. Is there nothing sort of in between "reflect this attribute no matter what" and "never ever reflect this attribute"?

```html
<component-datepicker
  range
  value="2025-01-12,2025-01-19"
  min="2025-01-01"
  max="2025-01-31"
  placeholder=""
  minlength="0"
  maxlength="4"
  pattern=""
  calendaropen>
```

Consider this element. `placeholder` is empty, which means it is displayed in the element as an empty string attribute. Same with `pattern`. And `minlength` and `maxlength` are only necessary if the type of datepicker is `multiple` and not `range`. Why do I have to show the default values for these in the element even though they are completely irrelevant for the element?

```html
<component-inputwrapper
  class="component-datepicker"
  label="Range"
  usewrapper
  helptext=""
  helptextdropdown=""
  countercurrent="2"
  countermaxlength="4"
  optionaltext="Valgfritt"
  requiredtext="Må fylles ut"
  errormessage="">
```

Same with this one. `helptext` and `helptextdropdown` are empty but they are always there, and `countercurrent` and `countermaxlength` are only relevant if the element has a counter, but I must always reflect their default values even when they are not used and irrelevant. I always show `optionalText` and `requiredText` as attributes because they have default values, even though `optional` and `required` boolean attributes are not set.

Are there any ways to gain better control of which attributes are reflected and when they are reflected? Feels like I'm placing a burden of messy DOM elements on the users of our components.

What I wish I could do was to have some attributes that are only reflected if they are set explicitly, so I don't have to expose the default values every time, and also have some attributes that are only reflected if certain criteria are met.

Is this possible in some way I have yet to learn, or do I just have to live with exposing empty, default and unused attributes to the DOM?

**Comment by tbroyer** [comment, marked as answer]
See https://github.com/lit/lit/issues/1476

**Comment by schjetne** [comment]
Thank you! So good times coming then! I'll just accept the messy DOM while we wait for this exciting new feature.
